src/jq_port.erl

```diff
@@ -0,0 +1,368 @@
+%% @doc gen_server callback module implementing a server controlling a jq port
+%% program
+%% @end
+%%%-------------------------------------------------------------------
+
+
+-module(jq_port).
+
+-behavior(gen_server).
+
+%% Gen server calls
+-export([
+          handle_call/3
+        , handle_cast/2
+        , init/1
+        , terminate/2
+        , handle_info/2
+        , code_change/3
+        ]).
+
+%% For supervisor
+
+-export([start_link/1]).
+
+%% Public interface
+
+-export([process_json/2]).
+
+%% Configuration functions
+-export([
+           set_filter_program_lru_cache_max_size/1
+         , get_filter_program_lru_cache_max_size/0
+         , nr_of_jq_port_servers/0
+         , set_nr_of_jq_port_servers/1
+        ]).
+
+%% Debug functionality
+
+-export([
+           start_recording/1
+         , stop_recording/0
+        ]).
+
+-define(APPNAME, jq).
+
+start_link(Id) ->
+    gen_server:start_link(jq_port, Id, []).
+
+do_op_ensure_started_helper(_Op, 0) ->
+    error("Could not start jq port servers");
+do_op_ensure_started_helper(Op, StartJQRetriesLeft) ->
+    try
+        Op()
+    catch
+        exit:{noproc, _} ->
+            application:stop(jq),
+            ok = application:ensure_started(jq),
+            do_op_ensure_started_helper(Op, StartJQRetriesLeft - 1)
+    end.
+
+do_op_ensure_started(Op) ->
+    do_op_ensure_started_helper(Op, 2).
+
+start_recording(FileName) ->
+    Op =
+    fun() -> 
+            Expect = [ok || _ <- lists:seq(0, jq_port:nr_of_jq_port_servers() - 1)],
+            Expect = [gen_server:call(port_server_by_id(Id),
+                                      {start_recording,
+                                       erlang:iolist_to_binary([FileName, "\0"])}) ||
+                      Id <- lists:seq(0, jq_port:nr_of_jq_port_servers() - 1)],
+            ok
+    end,
+    do_op_ensure_started(Op).
+
+
+stop_recording() ->
+    Op =
+    fun() ->
+            Expect = [ok || _ <- lists:seq(0, jq_port:nr_of_jq_port_servers() - 1)],
+            Expect = [gen_server:call(port_server_by_id(Id),
+                                      stop_recording) ||
+                      Id <- lists:seq(0, jq_port:nr_of_jq_port_servers() - 1)],
+            ok
+    end,
+    do_op_ensure_started(Op).
+
+get_filter_program_lru_cache_max_size() ->
+    Op =
+    fun() ->
+            gen_server:call(port_server(),
+                            get_filter_program_lru_cache_max_size)
+    end,
+    do_op_ensure_started(Op).
+
+set_filter_program_lru_cache_max_size(NewSize)
+  when is_integer(NewSize), NewSize >= 0, NewSize < 1073741824 ->
+    Op =
+    fun() ->
+            gen_server:call(port_server(),
+                            {set_filter_program_lru_cache_max_size, NewSize})
+    end,
+    do_op_ensure_started(Op).
+
+process_json(FilterProgram, JSONText)
+  when is_binary(FilterProgram), is_binary(JSONText) ->
+    Op =
+    fun() ->
+            gen_server:call(port_server(),
```

Comment:
infinity timeout.

Reply:
Good catch. Fixed in [1fd613b](https://github.com/emqx/jq/pull/18/commits/1fd613b11f7bee25f2c6cf0029c3717ddcb2d7e1).